subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -173,27 +191,42 @@ private String getFilename(Path dir) {
 
         @Override
         public FileVisitResult visitFile(Path file, @Nullable BasicFileAttributes attrs) {
-            String name = stringInterner.intern(file.getFileName().toString());
-            if (isAllowed(file, name, false, attrs, builder.getRelativePath())) {
+            String name = intern(file.getFileName().toString());
+            if (shouldVisit(file, name, false, attrs, builder.getRelativePath())) {
                 if (attrs == null) {
                     throw new FileSnapshottingException(String.format("Cannot read file '%s': not authorized.", file));
```

Comment:
Isn't this the case for an unreadable file? I thought we would add an `Unavailable` snapshot here.

Reply:
@lptr requested this explicitly, since this could mean that the parent folder is not readable.

Reply:
Can you document somewhere in the class where we end up when a file is unreadable? Given your comments it seems like:
- parent directory unreadable: `attrs == null` (-> Do we have a test for this?)
- file unreadable (`attrs.isRegularFile`) -> fails when trying to hash file
- directory unreadable (`!attrs.isRegularFile`) (is this true?) -> does not try to hash file
- when do we end up in `visitFileFailed`?

Reply:
I don't know how to reproduce `attrs == null` here, it's why I deleted it originally, since it didn't seem relevant to me after the change.
@lptr could you tell us an example?

> directory unreadable (!attrs.isRegularFile) (is this true?) -> does not try to hash file

> when do we end up in visitFileFailed?

This will actually call the above `visitFailed` for unlistable directories, it won't try to snapshot them

Reply:
Added comment to the unreadable dir snapshot and renamed the file snapshot method to reflect that it doesn't cover directories

Reply:
I would love to have an overview at the top of the class which unreadable files end up where. I don't think the comment in the method is enough. I am not even sure that it is true - i.e. I am not sure that `visitFileFailed` is only called for unreadable directories.

I am also afraid that by changing `isDirectory` from `false` to `true` in `visitFileFailed` we will break the use-case of filtering out unreadable directories in the input.

Reply:
Let's discuss it during daily, as I'm not sure what is expected of me here